devito/finite_differences/finite_difference.py

```diff
@@ -280,11 +280,19 @@ def indices_weights_to_fd(expr, dim, inds, weights, matvec=1):
             iloc = sympify(i).xreplace(mapper)
         # Enforce fixed precision FD coefficients to avoid variations in results
         c = sympify(c).evalf(_PRECISION)
-        terms.append(expr._subs(dim, iloc - (expr.indices_ref[dim] - dim)) * c)
 
-    deriv = EvalDerivative(*terms, base=expr)
+        # The FD term
+        term = expr._subs(dim, iloc - (expr.indices_ref[dim] - dim)) * c
+
+        # Re-evaluate any off-the-grid Functions potentially impacted by the FD
+        try:
+            term = term.evaluate
+        except AttributeError:
+            # Pure number
+            pass
+
+        terms.append(term)
```

Comment:
you could probably postponn the `*c` here as well, may be slightly more efficient (one less `Mul` built and evaluated). This also prevents the full travers of EvalDerivative so good catch

Reply:
Good catch. Note that this horribly conflicts with the changes in #1875 😬  so I'm going to rebase #1875 now on top of this one and will make the change there; I'll tag you there so that later you can give me green light here

Reply:
I tried, but I'm getting (expected) unevaluation errors:

```
self = <test_derivatives.TestFD object at 0x109a5dfa0>

    def test_fd_new_lo(self):
        grid = Grid((10,))
        x = grid.dimensions[0]
        u = Function(name="u", grid=grid, space_order=2)

        dplus = "-1.0*u(x)/h_x + u(x + 1.0*h_x)/h_x"
        dminus = "u(x)/h_x - 1.0*u(x - 1.0*h_x)/h_x"
>       assert str(u.dx(x0=x + .5 * x.spacing).evaluate) == dplus
E       AssertionError: assert '-1.0*u(x)/h_... 1.0*h_x)/h_x' == '-1.0*u(x)/h_... 1.0*h_x)/h_x'
E         - -1.0*u(x)/h_x + u(x + 1.0*h_x)/h_x
E         + -1.0*u(x)/h_x + 1.0*u(x + 1.0*h_x)/h_x
E         ?                 ++++

test_derivatives.py:368: AssertionError
```

Reply:
Looks like it's introducing a weird `1.0*` for some reason sympy being annoying again